### PR TITLE
Documentation related to timer

### PR DESCRIPTION
**Pull request by HyeJi**
Documentation related to timer

### CODE REVIEW

**Comment by kesl** [comment]
Can one of the admins verify this patch?

File: hypervisor/generic_timer/generic_timer.c [hunk 1]
```diff
@@ -32,6 +31,11 @@ hvmm_status_t generic_timer_set_tval(enum generic_timer_type type,
     return result;
 }
 
+/**
+* @brief Enable generic timer
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
+*/
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:38]: Remove this comment. Don't add comment in *.c file, except to static function.

File: hypervisor/generic_timer/generic_timer.c [hunk 2]
```diff
@@ -46,6 +50,11 @@ hvmm_status_t generic_timer_enable_int(enum generic_timer_type type)
     return result;
 }
 
+/**
+* @brief Disable generic timer
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
+*/
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:57]: Remove this comment. Don't add comment in *.c file, except to static function.

File: hypervisor/generic_timer/generic_timer.c [hunk 3]
```diff
@@ -60,11 +69,21 @@ hvmm_status_t generic_timer_disable_int(enum generic_timer_type type)
     return result;
 }
 
+/**
+* @brief Callback generic timer(hypervisor mode)
+* @param int irq, void *regs, void *pdata
+* @return void
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:75]: Do not add comment at void. Remove @return...

File: hypervisor/generic_timer/generic_timer.c [hunk 4]
```diff
@@ -60,11 +69,21 @@ hvmm_status_t generic_timer_disable_int(enum generic_timer_type type)
     return result;
 }
 
+/**
+* @brief Callback generic timer(hypervisor mode)
+* @param int irq, void *regs, void *pdata
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:74]: Please add comment about param information. It is just param's name. Like this..
@param irq Number of irq. 
@param *regs  Registers of context when an IRQ exception is raised.
@param *pdata Additional value. Not used this function.

File: hypervisor/generic_timer/generic_timer.c [hunk 5]
```diff
@@ -60,11 +69,21 @@ hvmm_status_t generic_timer_disable_int(enum generic_timer_type type)
     return result;
 }
 
+/**
+* @brief Callback generic timer(hypervisor mode)
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:73]: What is mean?
It is timer handler. Input comment information, When this handler is called.

File: hypervisor/generic_timer/generic_timer.c [hunk 6]
```diff
@@ -60,11 +69,21 @@ hvmm_status_t generic_timer_disable_int(enum generic_timer_type type)
...
 {
     _callback[GENERIC_TIMER_HYP](regs);
 }
 
+/**
+* @brief Enable generic timer irq
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
+*/
 hvmm_status_t generic_timer_enable_irq(enum generic_timer_type type)
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:87]: Remove this comment. Don't add comment in *.c file, except to static function.

File: hypervisor/generic_timer/generic_timer.c [hunk 7]
```diff
@@ -80,6 +99,11 @@ hvmm_status_t generic_timer_enable_irq(enum generic_timer_type type)
     return result;
 }
 
+/**
+* @brief Callback for setting
+* @param enum generic_timer_type type, generic_timer_callback_t callback
+* @return HVMM_STATUS_SUCCESS
+*/
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer.c:106]: Remove this comment. Don't add comment in *.c file, except to static function.

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 8]
```diff
@@ -27,6 +27,11 @@ enum {
     GENERIC_TIMER_REG_VIRT_OFF,
 };
 
+/**
+* @brief Generic timer write register value
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:31]: You mean, "write value to generic timer's register"?
Please, write to fit the meaning.

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 9]
```diff
@@ -27,6 +27,11 @@ enum {
...
+* @param : register, unit32_t
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:32]: what is register? what is uint32_t?
Please, write to fit the meaning.  Like this.
@param regiser register of generic timer.
@param val  Input value for generic timer's register.

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 10]
```diff
@@ -27,6 +27,11 @@ enum {
...
+* @return void
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:33]: Do not add comment about void

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 11]
```diff
@@ -64,6 +69,11 @@ static inline void generic_timer_reg_write(int reg, uint32_t val)
     isb();
 }
 
+/**
+* @brief Generic timer read register
+* @param int reg
+* @return uint32_t
+*/
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:76]: Refer to above

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 12]
```diff
@@ -102,6 +112,11 @@ static inline uint32_t generic_timer_reg_read(int reg)
     return val;
 }
 
+/**
+* @brief Generic timer write register(64bit)
+* @param int reg, uint64_t val
+* @return void
+*/
 static inline void generic_timer_reg_write64(int reg, uint64_t val)
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:120]: Refer to above

File: hypervisor/generic_timer/generic_timer_regs.h [hunk 13]
```diff
@@ -124,6 +139,11 @@ static inline void generic_timer_reg_write64(int reg, uint64_t val)
     isb();
 }
 
+/**
+* @brief Generic timer read register(64bit)
+* @param int reg
+* @return uint64_t
+*/
```
Comment (in1004kyu) [inline hypervisor/generic_timer/generic_timer_regs.h:146]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 14]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
 
 typedef void (*generic_timer_callback_t)(void *pdata);
 
-/* Calling this function is required only once in the entire system. */
+/**
+* @brief Numbering generic timer each level : HYP, NSP, VIR
+* Calling this function is required only once in the entire system.
+* @return HVMM_STATUS_SUCCESS (It takes "0")
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:21]: @return Result status.

File: hypervisor/include/generic_timer.h [hunk 15]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
 
 typedef void (*generic_timer_callback_t)(void *pdata);
 
-/* Calling this function is required only once in the entire system. */
+/**
+* @brief Numbering generic timer each level : HYP, NSP, VIR
+* Calling this function is required only once in the entire system.
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:20]: Press Enter

File: hypervisor/include/generic_timer.h [hunk 16]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
 
 typedef void (*generic_timer_callback_t)(void *pdata);
 
-/* Calling this function is required only once in the entire system. */
+/**
+* @brief Numbering generic timer each level : HYP, NSP, VIR
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:19]: @brief Set Irq number of generic timer.

File: hypervisor/include/generic_timer.h [hunk 17]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* Calling this function is required only once in the entire system.
+* @return HVMM_STATUS_SUCCESS (It takes "0")
+*/
 hvmm_status_t generic_timer_init();
-/* Enable the timer interrupt. Specified by timer type */
+
+/**
+* @brief Enable timer interrupt. Specified by timer type
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:28]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 18]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @brief Numbering generic timer each level : HYP, NSP, VIR
+* Calling this function is required only once in the entire system.
+* @return HVMM_STATUS_SUCCESS (It takes "0")
+*/
 hvmm_status_t generic_timer_init();
-/* Enable the timer interrupt. Specified by timer type */
+
+/**
+* @brief Enable timer interrupt. Specified by timer type
+* @param enum generic_timer_type type
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:27]: What type? Add comment about What it is

File: hypervisor/include/generic_timer.h [hunk 19]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
+*/
 hvmm_status_t generic_timer_enable_int(enum generic_timer_type type);
-/* Disable the timer. Specified by timer type */
+
+/**
+* @brief Disable the timer. Specified by timer type
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:35]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 20]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @brief Enable timer interrupt. Specified by timer type
+* @param enum generic_timer_type type
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
+*/
 hvmm_status_t generic_timer_enable_int(enum generic_timer_type type);
-/* Disable the timer. Specified by timer type */
+
+/**
+* @brief Disable the timer. Specified by timer type
+* @param enum generic_timer_type type
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:34]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 21]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+
 /*
  * Sets time interval. Converts from microseconds
  * to count and sets time interval.
  */
+
+
+/**
+* @brief Setting generic timer value
+* @param enum generic_timer_type type, uint32_t tval
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:47]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 22]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @return HVMM_STATUS_SUCCESS or HVMM_STATUS_UNSUPPORTED_FEATURE
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:48]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 23]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+*/
 hvmm_status_t generic_timer_set_tval(enum generic_timer_type type,
                     uint32_t tval);
-/* Enables timer irq.  */
+
+/**
+* @brief Enable generic timer irq
+* @param enum generic_timer_type type
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:55]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 24]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @return HVMM_STATUS_UNSUPPORTED_FEATURE or HVMM_STATUS_SUCCESS
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:56]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 25]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+*/
 hvmm_status_t generic_timer_enable_irq(enum generic_timer_type type);
+
 /* Adds callback funtion. Called when occur timer interrupt */
+
+/**
+* @brief Adds callback funtion. Called when occur timer interrupt
+* @param enum generic_timer_type type, generic_timer_callback_t callback
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:64]: Refer to above

File: hypervisor/include/generic_timer.h [hunk 26]
```diff
@@ -15,21 +15,55 @@ enum generic_timer_type {
...
+* @return HVMM_STATUS_SUCCESS
```
Comment (in1004kyu) [inline hypervisor/include/generic_timer.h:65]: Refer to above

File: hypervisor/sched_policy.h [hunk 27]
```diff
@@ -3,6 +3,11 @@
 
 #include <hvmm_types.h>
 
+/**
+*@brief It is schedule policy : determ next vm
+*@param void
```
Comment (in1004kyu) [inline hypervisor/sched_policy.h:8]: Do not comment about void

File: hypervisor/sched_policy.h [hunk 28]
```diff
@@ -3,6 +3,11 @@
...
+*@return vmid
```
Comment (in1004kyu) [inline hypervisor/sched_policy.h:9]: Do not comment about void

File: hypervisor/sched_policy.h [hunk 29]
```diff
@@ -3,6 +3,11 @@
 
 #include <hvmm_types.h>
 
+/**
+*@brief It is schedule policy : determ next vm
```
Comment (in1004kyu) [inline hypervisor/sched_policy.h:7]: Please, write to fit the meaning. Like this. What is determ
Do not comment like "It is"

@brief Determine next vmid

File: hypervisor/scheduler.c [hunk 30]
```diff
@@ -8,6 +8,11 @@ void scheduler_schedule(void)
     context_switchto(sched_policy_determ_next());
 }
 
+/**
+*@brief Testing context switch on guest
+*@param void *pdata
+*@return void
+*/
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:15]: Refer to above

File: hypervisor/scheduler.c [hunk 31]
```diff
@@ -31,10 +36,16 @@ void scheduler_test_switch_to_next_guest(void *pdata)
         scheduler_schedule();
 }
 
+/**
+*@param void *pdata
+*@return void
+*@todo Have to implement
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:42]: No. It is extra timer callback function. When you need to additional timer callback function. Use this.

File: hypervisor/scheduler.c [hunk 32]
```diff
@@ -31,10 +36,16 @@ void scheduler_test_switch_to_next_guest(void *pdata)
         scheduler_schedule();
 }
 
+/**
+*@param void *pdata
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:40]: Data from timer handler. Typically Registers of context when an IRQ exception is raised.

File: hypervisor/scheduler.c [hunk 33]
```diff
@@ -31,10 +36,16 @@ void scheduler_test_switch_to_next_guest(void *pdata)
...
+*@return void
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:41]: REFER TO ABOVE

File: hypervisor/scheduler.c [hunk 34]
```diff
@@ -31,10 +36,16 @@ void scheduler_test_switch_to_next_guest(void *pdata)
...
+*@todo Have to implement
+*/
 void extra_timer_callback(void *pdata)
 {
 }
 
+
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:48]: Do not additional empty line.

File: hypervisor/scheduler.c [hunk 35]
```diff
@@ -31,10 +36,16 @@ void scheduler_test_switch_to_next_guest(void *pdata)
         scheduler_schedule();
 }
 
+/**
```
Comment (in1004kyu) [inline hypervisor/scheduler.c:39]: REFER TO ABOVE

File: hypervisor/scheduler.h [hunk 36]
```diff
@@ -7,16 +7,31 @@
 
 #include <log/uart_print.h>
 
+/**
+*@param int irq, void *pdata
+*@todo Have to implement
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:12]: This is implemented.
Check back

File: hypervisor/scheduler.h [hunk 37]
```diff
@@ -7,16 +7,31 @@
 
 #include <log/uart_print.h>
 
+/**
+*@param int irq, void *pdata
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:11]: REFER TO ABOVE.

File: hypervisor/scheduler.h [hunk 38]
```diff
@@ -7,16 +7,31 @@
...
+*@todo Have to implement
+*/
 hvmm_status_t scheduler_next_event(int irq, void *pdata);
 
-/* Test Code */
+/**
+*@brief Test scheduling
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:17]: No. REFER TO ABOVE. Read code. Do not guess just reading function name. It is need to rename function name.
Ask to me when you wonder function name.

File: hypervisor/scheduler.h [hunk 39]
```diff
@@ -7,16 +7,31 @@
...
+*@param void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:18]: REFER TO ABOVE.

File: hypervisor/scheduler.h [hunk 40]
```diff
@@ -7,16 +7,31 @@
...
+*@return void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:19]: REFER TO ABOVE.

File: hypervisor/scheduler.h [hunk 41]
```diff
@@ -7,16 +7,31 @@
...
+*/
 void scheduler_test_scheduling(void);
 
-/*
- * Schedules guest context switch according to the default
- * scheduling policy (sched_policy.c)
- */
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:18]: Why did you remove this comment? Do not remove this.

File: hypervisor/scheduler.h [hunk 42]
```diff
@@ -7,16 +7,31 @@
...
+/**
+*@brief Request switch when exit trap
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:24]: @brief Scheduling to next vm.

File: hypervisor/scheduler.h [hunk 43]
```diff
@@ -7,16 +7,31 @@
...
+*@param void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:25]: REFER TO ABOVE

File: hypervisor/scheduler.h [hunk 44]
```diff
@@ -7,16 +7,31 @@
...
+*@return void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:26]: REFER TO ABOVE

File: hypervisor/scheduler.h [hunk 45]
```diff
@@ -7,16 +7,31 @@
...
+*/
 void scheduler_schedule(void);
+
+/**
+*@brief Test context switch to next guest
+*@param void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:32]: Not void. void is type. param is *pdata.

File: hypervisor/scheduler.h [hunk 46]
```diff
@@ -7,16 +7,31 @@
...
+*@return void
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:33]: REFER TO ABOVE

File: hypervisor/scheduler.h [hunk 47]
```diff
@@ -7,16 +7,31 @@
...
- */
+/**
+*@brief Request switch when exit trap
+*@param void
+*@return void
+*/
 void scheduler_schedule(void);
+
+/**
+*@brief Test context switch to next guest
```
Comment (in1004kyu) [inline hypervisor/scheduler.h:31]: @brief Determine Whether scheduler call or not.

File: hypervisor/timer.c [hunk 48]
```diff
@@ -12,6 +12,11 @@
 
 static struct timer_channel _channels[TIMER_NUM_MAX_CHANNELS];
 
+/**
+* @brief callback timer channel
+* @param enum timer_channel_type channel, void *param
+* @return void
```
Comment (in1004kyu) [inline hypervisor/timer.c:18]: REFER TO ABOVE.

File: hypervisor/timer.c [hunk 49]
```diff
@@ -12,6 +12,11 @@
 
 static struct timer_channel _channels[TIMER_NUM_MAX_CHANNELS];
 
+/**
+* @brief callback timer channel
```
Comment (in1004kyu) [inline hypervisor/timer.c:16]: @brief Call all callback function of same timer channel.

File: hypervisor/timer.c [hunk 50]
```diff
@@ -12,6 +12,11 @@
...
+* @param enum timer_channel_type channel, void *param
```
Comment (in1004kyu) [inline hypervisor/timer.c:17]: Please, write to fit the meaning.

File: hypervisor/timer.c [hunk 51]
```diff
@@ -21,6 +26,11 @@ static void _timer_each_callback_channel(enum timer_channel_type channel,
             _channels[channel].callbacks[i](param);
 }
 
+/**
+* @brief count timer channel callback
```
Comment (in1004kyu) [inline hypervisor/timer.c:30]: Please, write to fit the meaning. What is counter timer channel callback?

File: hypervisor/timer.c [hunk 52]
```diff
@@ -21,6 +26,11 @@ static void _timer_each_callback_channel(enum timer_channel_type channel,
...
+* @param  enum timer_channel_type channel
```
Comment (in1004kyu) [inline hypervisor/timer.c:31]: Please, write to fit the meaning.

File: hypervisor/timer.c [hunk 53]
```diff
@@ -21,6 +26,11 @@ static void _timer_each_callback_channel(enum timer_channel_type channel,
...
+* @return int count value
```
Comment (in1004kyu) [inline hypervisor/timer.c:32]: Please, write to fit the meaning.

File: hypervisor/timer.c [hunk 54]
```diff
@@ -30,6 +40,11 @@ static int _timer_channel_num_callbacks(enum timer_channel_type channel)
     return count;
 }
 
+/**
+* @brief callback HW
```
Comment (in1004kyu) [inline hypervisor/timer.c:44]: Please, write to fit the meaning.

File: hypervisor/timer.c [hunk 55]
```diff
@@ -30,6 +40,11 @@ static int _timer_channel_num_callbacks(enum timer_channel_type channel)
...
+* @param void *pdata
```
Comment (in1004kyu) [inline hypervisor/timer.c:45]: Please, write to fit the meaning.

File: hypervisor/timer.c [hunk 56]
```diff
@@ -30,6 +40,11 @@ static int _timer_channel_num_callbacks(enum timer_channel_type channel)
...
+* @return void
```
Comment (in1004kyu) [inline hypervisor/timer.c:46]: REMOVE

File: hypervisor/timer.h [hunk 57]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
     timer_callback_t callbacks[TIMER_MAX_CHANNEL_CALLBACKS];
 };
 
-/*
- * Calling this function is required only once in the entire system
- * prior to calls to other functions of Timer module.
- */
+/**
+* @brief Calling this function is required only once in the entire system
+* prior to calls<br>to other functions of Timer module.
```
Comment (in1004kyu) [inline hypervisor/timer.h:45]: PRESS ENTER. 
Why did you insert &lt;br&gt;?

File: hypervisor/timer.h [hunk 58]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
     timer_callback_t callbacks[TIMER_MAX_CHANNEL_CALLBACKS];
 };
 
-/*
- * Calling this function is required only once in the entire system
- * prior to calls to other functions of Timer module.
- */
+/**
```
Comment (in1004kyu) [inline hypervisor/timer.h:43]: @brief Initialize the timer module.

File: hypervisor/timer.h [hunk 59]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @brief Calling this function is required only once in the entire system
+* prior to calls<br>to other functions of Timer module.
+* @param enum timer_channel_t channel
```
Comment (in1004kyu) [inline hypervisor/timer.h:46]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 60]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_SUCCES (It takes '0')
```
Comment (in1004kyu) [inline hypervisor/timer.h:47]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 61]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
- * Starts the timer channel specified by 'channel'. The callback,
- * if set, will be periodically called until it's unset or the channel stops by
- * 'timer_stop(timer_channel)'
- */
+/**
+* @brief
+* Starts the timer channel specified by 'channel'.
+* The callback, if set, will be periodically called
+* until it's unset or the channel stops
+* by 'timer_stop(timer_channel)i'
```
Comment (in1004kyu) [inline hypervisor/timer.h:56]: remove i

File: hypervisor/timer.h [hunk 62]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @param enum timer_channel_type channel
```
Comment (in1004kyu) [inline hypervisor/timer.h:57]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 63]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_SUCCES (It takes '0')
```
Comment (in1004kyu) [inline hypervisor/timer.h:58]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 64]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+*/
 hvmm_status_t timer_start(enum timer_channel_type channel);
 
-/*
- *  Stops the timer channel specified by 'channel'
- */
+/**
+* @brief
+* Stops the timer channel specified by 'channel'
+* @param enum timer_channel_t channel
```
Comment (in1004kyu) [inline hypervisor/timer.h:65]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 65]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_SUCCES (It takes '0')
```
Comment (in1004kyu) [inline hypervisor/timer.h:66]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 66]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
- * Sets time interval, in microseconds, for the timer channel.
- * If the channel has been started and a callback function is set,
- * it will be called in the next interval
- */
+/**
+* @brief
+* Sets time interval, in microseconds, for the timer channel.
+* If the channel has been started and a callback function is set,
+* it will be called in the next interval
+* @param enum timer_channel_t channel, uint32_t interval_us
```
Comment (in1004kyu) [inline hypervisor/timer.h:75]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 67]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_SUCCES (It takes '0')
```
Comment (in1004kyu) [inline hypervisor/timer.h:76]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 68]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
 
-/*
- * Returns the time interval for the timer channel if it was set previously.
- * Unknown value is returned otherwise.
- */
+/**
+* @brief
+* Returns the time interval for the timer channel if it was set previously.
+* Unknown value is returned otherwise.
+* @param enum timer_channel_type channel
```
Comment (in1004kyu) [inline hypervisor/timer.h:85]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 69]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return uint32_t _channels
```
Comment (in1004kyu) [inline hypervisor/timer.h:86]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 70]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+*/
 uint32_t timer_get_interval(enum timer_channel_type channel);
 
-/*
- * Adds a callback function for the timer channel.
- */
+/**
+* @brief
+* Adds a callback function for the timer channel.
+* @param enum timer_channel_t channel, timer_callback_t callback
```
Comment (in1004kyu) [inline hypervisor/timer.h:93]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 71]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_BUSY or HVMM_STATUS_SUCESS
```
Comment (in1004kyu) [inline hypervisor/timer.h:94]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 72]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
- * registered callback function list
- * if it previously has been added.
- */
-hvmm_status_t timer_remove_callback(enum timer_channel_type channel,
+/**
+* @brief
+* Removes the callback function from the timer channel's
+* registered callback function list
+* if it previously has been added.
+* @param timer_channel_t channel, timer_callback_t callback
```
Comment (in1004kyu) [inline hypervisor/timer.h:104]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 73]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return hvmm_status_t HVMM_STATUS_NOT_FOUND or HVMM_STATUS_SUCESS
```
Comment (in1004kyu) [inline hypervisor/timer.h:105]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 74]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+hvmm_status_t enum timer_remove_callback(enum timer_channel_type channel,
         timer_callback_t handler);
 
-/*
- * Converts from microseconds to system counter.
- */
+/**
+* @brief
+* Converts from microseconds to system counter.
+* @param uint64_t time_us
```
Comment (in1004kyu) [inline hypervisor/timer.h:113]: Please, write to fit the meaning.

File: hypervisor/timer.h [hunk 75]
```diff
@@ -40,55 +40,79 @@ struct timer_channel {
...
+* @return uint64_t time_us * COUNT_PER_USEC
```
Comment (in1004kyu) [inline hypervisor/timer.h:114]: Please, write to fit the meaning.

**Comment by in1004kyu** [comment]
Please, write to fit the meaning. And read code.

**Comment by Wonseok** [comment]
This is not a documentation. Please fix the title such as "Added doxygen style comments for timer"

File: hypervisor/generic_timer/generic_timer.c [hunk 76]
```diff
@@ -32,6 +31,11 @@ hvmm_status_t generic_timer_set_tval(enum generic_timer_type type,
     return result;
 }
```
Comment (Wonseok) [inline hypervisor/generic_timer/generic_timer.c:32]: 1. In else case we do not handle anything. Why do you leave it?
2. In this function do not need result variable isn't it?

You might want to fix it as below.
      if (type == GENERIC_TIMER_HYP) {
          _tvals[type] = tval;
          generic_timer_reg_write(GENERIC_TIMER_REG_HYP_TVAL, tval);
          return HVMM_STATUS_SUCCESS;
      } else {
          return HVMM_STATUS_UNSUPPORTED_FEATURE;
      }
